src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,179 @@
+package duke;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+import command.Command;
+import command.DeadlineCommand;
+import command.DeleteCommand;
+import command.EventCommand;
+import command.FindCommand;
+import command.ListCommand;
+import command.MarkCommand;
+import command.TodoCommand;
+import command.UnmarkCommand;
+
+/**
+ * Parses user input
+ */
+public class Parser {
+    private static final DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+
+    /**
+     * Parses user input into command for execution.
+     *
+     * @param input user input string
+     * @return the command based on the user input
+     */
+    public static Command parse(String input) throws DukeException {
+        if (input.equals("list")) {
+            return new ListCommand();
+        } else {
+            String[] details = input.split(" ", 2);
+            String commandName = details[0];
+            switch (commandName) {
+            case MarkCommand.COMMAND_WORD:
+                return prepareMark(details);
+            case UnmarkCommand.COMMAND_WORD:
+                return prepareUnmark(details);
+            case DeleteCommand.COMMAND_WORD:
+                return prepareDelete(details);
+            case TodoCommand.COMMAND_WORD:
+                return prepareTodo(details);
+            case DeadlineCommand.COMMAND_WORD:
+                return prepareDeadline(details);
+            case EventCommand.COMMAND_WORD:
+                return prepareEvent(details);
+            case FindCommand.COMMAND_WORD:
+                return prepareFind(details);
+            default:
+                throw new DukeException("Sorry! I do not recognise this command");
+            }
+        }
+    }
+
+    /**
+     * Parses arguments in the context of the mark task command
+     *
+     * @param details full command args string
+     * @return the prepared command
+     */
+    private static MarkCommand prepareMark(String[] details) throws DukeException {
+        // user input only has the command eg "mark"
+        if (details.length < 2 || details[1].trim().isEmpty()) {
```

Comment:
The code has repeated checks for details length and emptiness. Consider creating a helper method to standardize these checks.
```java
private static void validateInputDetails(String[] details) throws DukeException {
    if (details.length < 2 || details[1].trim().isEmpty()) {
        throw new DukeException("Invalid command! Please provide more details");
    }
}
```